hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AggregateMetricsManager.java

```diff
@@ -0,0 +1,175 @@
+
+    if (dispatchIntervalInMins <= 0) {
+      throw new InvalidConfigurationValueException(
+          "dispatchIntervalInMins must be > 0");
+    }
+
+    if (permitsPerSecond <= 0) {
+      throw new InvalidConfigurationValueException(
+          "permitsPerSecond must be > 0");
+    }
+
+    this.rateLimiter = new SimpleRateLimiter(permitsPerSecond);
+
+    // Initialize scheduler for periodic dispatching of metrics.
+    this.scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
+      Thread t = new Thread(r, "ABFS-Aggregated-Metrics-Dispatcher");
+      t.setDaemon(true);
+      return t;
+    });
+
+    // Schedule periodic dispatching of metrics.
+    this.scheduler.scheduleAtFixedRate(
+        this::dispatchMetrics,
+        dispatchIntervalInMins,
+        dispatchIntervalInMins,
+        TimeUnit.MINUTES);
+
+    // Add shutdown hook to dispatch remaining metrics on JVM shutdown.
+    Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+      dispatchMetrics();
+      scheduler.shutdown();
+    }));
+  }
+
+  /**
+   * Get the singleton instance of AggregateMetricsManager.
+   *
+   * @param dispatchIntervalInMins Interval in minutes for dispatching metrics.
+   * @param permitsPerSecond       Rate limit for dispatching metrics.
+   * @return Singleton instance of AggregateMetricsManager.
+   */
+  public static AggregateMetricsManager get(final long dispatchIntervalInMins,
+      final int permitsPerSecond) {
+    if (instance != null) {
+      return instance;
+    }
+
+    synchronized (AggregateMetricsManager.class) {
+      if (instance == null) {
+        try {
+          instance = new AggregateMetricsManager(
+              dispatchIntervalInMins, permitsPerSecond);
+        } catch (InvalidConfigurationValueException e) {
+          throw new RuntimeException(
+              "Failed to initialize AggregateMetricsManager", e);
+        }
+      }
+      return instance;
+    }
+  }
+
+  /**
+   * Register an AbfsClient with the manager.
+   * @param account Account name.
+   * @param abfsClient AbfsClient instance.
+   */
+  public void registerClient(String account, AbfsClient abfsClient) {
+    if (StringUtils.isEmpty(account) || abfsClient == null) {
+      return;
+    }
+
+    buckets.computeIfAbsent(account,
+            key -> new MetricsBucket(rateLimiter))
+        .registerClient(abfsClient);
+  }
+
+  /**
+   * Deregister an AbfsClient from the manager.
+   * @param account Account name.
+   * @param abfsClient AbfsClient instance.
+   * @return true if the client was deregistered, false otherwise.
+   */
+  public boolean deregisterClient(String account, AbfsClient abfsClient) {
+    if (StringUtils.isEmpty(account) || abfsClient == null) {
+      return false;
+    }
+
+    MetricsBucket bucket = buckets.get(account);
+    if (bucket == null) {
+      return false;
+    }
+
+    boolean isRemoved = bucket.deregisterClient(abfsClient);
+
+    if (bucket.isEmpty()) {
```

Comment:
There is a small race window between isEmpty() and remove(). Another thread may concurrently register a new client for the same account and reuse the bucket, but it can still be removed based on the earlier emptiness check. This makes the behavior timing-dependent and hard to reason about under concurrency.

You can use buckets.computeIfPresent() to perform the emptiness check and removal atomically, which avoids this race and keeps the map state consistent.